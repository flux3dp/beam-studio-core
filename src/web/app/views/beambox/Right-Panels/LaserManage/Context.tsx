import React, { Dispatch } from 'react';

import i18n from 'helpers/i18n';
import isObjectEmpty from 'helpers/is-object-empty';
import storage from 'implementations/storage';
import { getDefaultPresetData } from 'helpers/presets/preset-helper';
import { ILaserConfig, ILaserDataChanges } from 'interfaces/ILaserConfig';

interface State {
  configs: ILaserConfig[];
  selectedItem : {
    name: string;
    isCustomized: boolean;
  };
  presetsInUse: { [name: string]: boolean };
  dataChanges: { [key: string]: ILaserDataChanges };
  displayValues: {
    speed: number;
    power: number;
    repeat: number;
    zStep: number;
  };
}

interface SelectPayload {
  name: string;
  isCustomized: boolean;
}
interface ChangePayload {
  name: string;
  key: 'speed' | 'power' | 'repeat' | 'zStep';
  value: number;
}
interface AddConfigPayload {
  name: string;
}
interface SwapConfigPayload {
  orig: number;
  dist: number;
}

type Action = {
  type: 'select';
  payload: SelectPayload;
} | {
  type: 'change';
  payload: ChangePayload;
} | {
  type: 'add-config';
  payload: AddConfigPayload;
} | {
  type: 'swap-config';
  payload: SwapConfigPayload;
} | {
  type: 'add-preset';
  payload?: never;
} | {
  type: 'remove-preset';
  payload?: never;
} | {
  type: 'delete';
  payload?: never;
} | {
  type: 'reset';
  payload?: never;
};

export const getInitState = (selectingConfigName: string): State => {
  const configs: ILaserConfig[] = storage.get('customizedLaserConfigs') || [];
  const presetsInUse = storage.get('defaultLaserConfigsInUse');
  const selectedConfig = configs.find((c) => c.name === selectingConfigName);
  const { speed, power, repeat, zStep } = selectedConfig || {};
  return {
    configs,
    selectedItem: { name: selectingConfigName, isCustomized: true },
    presetsInUse,
    dataChanges: {},
    displayValues: {
      speed: speed || 1,
      power: power || 0,
      repeat: repeat || 1,
      zStep: zStep || 0,
    },
  };
};

const selectAction = (state: State, payload: SelectPayload): State => {
  const { name, isCustomized } = payload;
  if (isCustomized) {
    const { configs, dataChanges } = state;
    const selectedConfig = configs.find((c) => c.name === name);
    const { speed, power, repeat, zStep } = { ...selectedConfig, ...dataChanges[name] };
    return {
      ...state,
      selectedItem: { name, isCustomized },
      displayValues: { speed, power, repeat, zStep },
    };
  }
  // selecting presets
  const { speed, power, repeat } = getDefaultPresetData(name);
  const zStep = 0;
  return {
    ...state,
    selectedItem: { name, isCustomized },
    displayValues: { speed, power, repeat, zStep },
  };
};

const changeAction = (state: State, payload: ChangePayload): State => {
  const { name, key, value } = payload;
  const { configs, dataChanges, displayValues } = state;
  const selectedConfig = configs.find((c) => c.name === name);
  if (selectedConfig[key] !== value) {
    if (!dataChanges[name]) {
      const change = {};
      change[key] = value;
      dataChanges[name] = change;
    } else {
      dataChanges[name][key] = value;
    }
  } else {
    if (dataChanges[name]) {
      delete dataChanges[name][key];
    }
    if (isObjectEmpty(dataChanges[name])) {
      delete dataChanges[name];
    }
  }
  return {
    ...state,
    dataChanges,
    displayValues: {
      ...displayValues,
      [key]: value,
    },
  };
};

const addConfigAction = (state: State, payload: AddConfigPayload): State => {
  const { name } = payload;
  const { configs } = state;
  configs.push({ name, speed: 20, power: 15, repeat: 1, zStep: 0 });
  return {
    ...state,
    selectedItem: { name, isCustomized: true },
    displayValues: { speed: 20, power: 15, repeat: 1, zStep: 0 },
  };
};

const swapConfigAction = (state: State, payload: SwapConfigPayload): State => {
  const { orig, dist } = payload;
  const { configs } = state;
  [configs[orig], configs[dist]] = [configs[dist], configs[orig]];
  return {
    ...state,
    configs,
  };
};

const addPresetAction = (state: State): State => {
  const { selectedItem: { name: key, isCustomized }, configs, presetsInUse } = state;
  if (!isCustomized && key !== '') {
    const unit = storage.get('default-units') || 'mm';
    const displayName = i18n.lang.beambox.right_panel.laser_panel.dropdown[unit][key];
    if (presetsInUse[key]) {
      return { ...state, selectedItem: { name: displayName, isCustomized: true } };
    }
    const { speed, power, repeat } = getDefaultPresetData(key);
    presetsInUse[key] = true;
    configs.push({ name: displayName, speed, power, repeat, zStep: 0, isDefault: true, key });
    return {
      ...state,
      selectedItem: { name: displayName, isCustomized: true },
      configs,
      presetsInUse,
    };
  }
  return state;
};

const removePresetAction = (state: State): State => {
  const { selectedItem: { name, isCustomized }, configs, presetsInUse, dataChanges } = state;
  const unit = storage.get('default-units') || 'mm';
  const configName = isCustomized ? name : i18n.lang.beambox.right_panel.laser_panel.dropdown[unit][name];
  const index = configs.findIndex((c) => c.name === configName);

  if (index < 0 || !configs[index].isDefault) return state;
  const { key } = configs[index];
  presetsInUse[key] = false;
  configs.splice(index, 1);
  if (configs.length > 0) {
    // if custom config list is not empty, selecting config above
    const newIndex = Math.min(index, configs.length - 1);
    const newConfig = configs[newIndex];
    const { name: newConfgName, speed, power, repeat, zStep } = { ...newConfig, ...dataChanges[newConfig.name] };
    return {
      ...state,
      configs,
      presetsInUse,
      selectedItem: { name: newConfgName, isCustomized: true },
      displayValues: { speed, power, repeat: repeat || 1, zStep: zStep || 0 },
    };
  }
  // selecting remove preset in preset list
  return {
    ...state,
    configs,
    presetsInUse,
    selectedItem: { name, isCustomized: false },
  };
};

const deleteAction = (state: State): State => {
  const { selectedItem: { name, isCustomized }, configs, dataChanges } = state;
  if (!isCustomized) return state;
  const idx = configs.findIndex((c) => c.name === name);
  if (idx < 0) return state;
  if (configs[idx].isDefault) return removePresetAction(state);
  configs.splice(idx, 1);
  if (configs.length > 0) {
    // if custom config list is not empty, selecting config above
    const newIndex = Math.min(idx, configs.length - 1);
    const newConfig = configs[newIndex];
    const { name: newConfgName, speed, power, repeat, zStep } = { ...newConfig, ...dataChanges[newConfig.name] };
    return {
      ...state,
      configs,
      selectedItem: { name: newConfgName, isCustomized: true },
      displayValues: { speed, power, repeat: repeat || 1, zStep: zStep || 0 },
    };
  }
  return {
    ...state,
    configs,
    selectedItem: { name: '', isCustomized: false },
    displayValues: { speed: 1, power: 0, repeat: 1, zStep: 0 },
  };
};

const resetAction = (state: State): State => {
  const configs = storage.get('customizedLaserConfigs') || [];
  const presetsInUse = storage.get('defaultLaserConfigsInUse') || {};
  return { ...state, configs, presetsInUse };
};

export const reducer = (state: State, action: Action): State => {
  const { type, payload } = action;
  switch (type) {
    case 'select':
      return selectAction(state, payload);
    case 'change':
      return changeAction(state, payload);
    case 'add-config':
      return addConfigAction(state, payload);
    case 'swap-config':
      return swapConfigAction(state, payload);
    case 'add-preset':
      return addPresetAction(state);
    case 'remove-preset':
      return removePresetAction(state);
    case 'delete':
      return deleteAction(state);
    case 'reset':
      return resetAction(state);
    default:
      break;
  }
  return state;
};

interface Context {
  state: State;
  dispatch: Dispatch<Action>;
}

export default React.createContext<Context>({} as Context);
