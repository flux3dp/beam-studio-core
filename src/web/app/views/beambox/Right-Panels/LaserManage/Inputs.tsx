import classNames from 'classnames';
import React, { memo, useContext, useMemo } from 'react';
import { Col } from 'antd';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';

import Context from './Context';
import styles from './Inputs.module.scss';

const Inputs = () => {
  const { dispatch, state } = useContext(Context);
  const { selectedItem: { name, isCustomized }, configs, displayValues } = state;
  const lang = useI18n();
  const t = lang.beambox.right_panel.laser_panel;

  const speedLimit = useMemo(() => {
    const model = BeamboxPreference.read('workarea') || BeamboxPreference.read('model');
    return model === 'fhexa1' ? 900 : 300;
  }, []);

  const unit = useMemo(() => storage.get('default-units') || 'mm', []);
  const { speedUnit, speedDecimal, zStepUnit, zStepDecimal, zStepStep } = useMemo(() => ({
    speedUnit: { mm: 'mm/s', inches: 'in/s' }[unit],
    speedDecimal: { mm: 1, inches: 3 }[unit],
    zStepUnit: { mm: 'mm', inches: 'in' }[unit],
    zStepDecimal: { mm: 2, inches: 4 }[unit],
    zStepStep: { mm: 0.5, inches: 0.01 }[unit],
  }), [unit]);

  const selectedConfig = configs.find((c) => c.name === name);
  const disabled = !isCustomized || !selectedConfig || selectedConfig.isDefault;
  const { power, speed, repeat, zStep } = displayValues;

  return (
    <>
      <div>
        <br />
        <strong>
          {isCustomized ? name : t.dropdown[unit][name]}
        </strong>
      </div>
      <div className={classNames(styles.container, { [styles.disabled]: disabled })}>
        <Col span={11}>
          <div className={styles.control}>
            <span className={styles.label}>{t.power.text}</span>
            <UnitInput
              id="laser-power"
              className={{ [styles.input]: true }}
              min={1}
              max={100}
              disabled={disabled}
              unit="%"
              getValue={(value) => dispatch({ type: 'change', payload: { name, key: 'power', value } })}
              defaultValue={power}
              decimal={1}
              step={1}
            />
          </div>
          <div className={styles.control}>
            <span className={styles.label}>{t.laser_speed.text}</span>
            <UnitInput
              id="laser-speed"
              className={{ [styles.input]: true }}
              min={3}
              max={speedLimit}
              disabled={disabled}
              unit={speedUnit}
              getValue={(value) => dispatch({ type: 'change', payload: { name, key: 'speed', value } })}
              defaultValue={speed}
              decimal={speedDecimal}
              step={1}
            />
          </div>
        </Col>
        <Col span={11}>
          <div className={styles.control}>
            <span className={styles.label}>{t.repeat}</span>
            <UnitInput
              id="laser-repeat"
              className={{ [styles.input]: true }}
              min={1}
              max={100}
              disabled={disabled}
              unit={t.times}
              getValue={(value) => dispatch({ type: 'change', payload: { name, key: 'repeat', value } })}
              defaultValue={repeat}
              decimal={0}
              step={1}
            />
          </div>
          <div className={styles.control}>
            <span className={styles.label}>{t.z_step}</span>
            <UnitInput
              id="laser-z-step"
              className={{ [styles.input]: true }}
              min={0}
              max={20}
              disabled={disabled}
              unit={zStepUnit}
              getValue={(value) => dispatch({ type: 'change', payload: { name, key: 'zStep', value } })}
              defaultValue={zStep}
              decimal={zStepDecimal}
              step={zStepStep}
            />
          </div>
        </Col>
      </div>
    </>
  );
};

export default memo(Inputs);
