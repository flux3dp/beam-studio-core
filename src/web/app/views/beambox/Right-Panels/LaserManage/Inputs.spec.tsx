import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import Context from './Context';
import Inputs from './Inputs';

jest.mock('app/views/beambox/Right-Panels/LaserManage/Context', () => React.createContext({}));

const mockRead = jest.fn();
jest.mock('app/actions/beambox/beambox-preference', () => ({
  read: (...args) => mockRead(...args),
}));

const mockGet = jest.fn();
jest.mock('implementations/storage', () => ({
  get: (...args) => mockGet(...args),
}));

jest.mock('helpers/useI18n', () => () => ({
  beambox: {
    right_panel: {
      laser_panel: {
        repeat: 'repeat',
        z_step: 'z_step',
        power: { text: 'power' },
        laser_speed: { text: 'laser_speed' },
        dropdown: {
          mm: {
            preset1: 'preset1',
          },
        },
      },
    },
  },
}));

jest.mock('app/widgets/Unit-Input-v2', () => ({
  id, min, max, unit, decimal, step, defaultValue, getValue,
}: any) => (
  <div>
    Dummy Unit Input
    min: {min}
    max: {max}
    unit: {unit}
    defaultValue: {defaultValue}
    decimal: {decimal}
    step: {step}
    <button type="button" onClick={() => getValue(7)}>{id}</button>
  </div>
));

const mockDispatch = jest.fn();
describe('test PresetsList', () => {
  beforeEach(() => {
    mockGet.mockReturnValue('mm');
    mockRead.mockReturnValue('fhex1');
  });

  it('should render correctly', () => {
    const mockState = {
      selectedItem: { name: 'preset1', isCustomized: true },
      configs: [{ name: 'preset1', isDefault: false }],
      displayValues: { power: 5, speed: 5, repeat: 5, zStep: 5 },
    };
    const { container } = render(
      <Context.Provider value={{ state: mockState as any, dispatch: mockDispatch }}>
        <Inputs />
      </Context.Provider>
    );
    expect(container).toMatchSnapshot();
    expect(mockGet).toBeCalledTimes(1);
    expect(mockGet).toHaveBeenLastCalledWith('default-units');
    expect(mockRead).toBeCalledTimes(1);
    expect(mockRead).toHaveBeenLastCalledWith('workarea');
  });

  test('events should work', () => {
    const mockState = {
      selectedItem: { name: 'preset1', isCustomized: true },
      configs: [{ name: 'preset1', isDefault: false }],
      displayValues: { power: 5, speed: 5, repeat: 5, zStep: 5 },
    };
    const { getByText } = render(
      <Context.Provider value={{ state: mockState as any, dispatch: mockDispatch }}>
        <Inputs />
      </Context.Provider>
    );
    fireEvent.click(getByText('laser-power'));
    expect(mockDispatch).toBeCalledTimes(1);
    expect(mockDispatch).toHaveBeenLastCalledWith(
      { type: 'change', payload: { name: 'preset1', key: 'power', value: 7 } }
    );

    fireEvent.click(getByText('laser-speed'));
    expect(mockDispatch).toBeCalledTimes(2);
    expect(mockDispatch).toHaveBeenLastCalledWith(
      { type: 'change', payload: { name: 'preset1', key: 'speed', value: 7 } }
    );

    fireEvent.click(getByText('laser-repeat'));
    expect(mockDispatch).toBeCalledTimes(3);
    expect(mockDispatch).toHaveBeenLastCalledWith(
      { type: 'change', payload: { name: 'preset1', key: 'repeat', value: 7 } }
    );

    fireEvent.click(getByText('laser-z-step'));
    expect(mockDispatch).toBeCalledTimes(4);
    expect(mockDispatch).toHaveBeenLastCalledWith(
      { type: 'change', payload: { name: 'preset1', key: 'zStep', value: 7 } }
    );
  });
});
