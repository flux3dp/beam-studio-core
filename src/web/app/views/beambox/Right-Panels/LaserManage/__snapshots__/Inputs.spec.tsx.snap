// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PresetsList should render correctly 1`] = `
<div>
  <div>
    <br />
    <strong>
      preset1
    </strong>
  </div>
  <div
    class="container"
  >
    <div
      class="ant-col ant-col-11 css-dev-only-do-not-override-hash"
    >
      <div
        class="control"
      >
        <span
          class="label"
        >
          power
        </span>
        <div>
          Dummy Unit Input min: 
          1
          max: 
          100
          unit: 
          %
          defaultValue: 
          5
          decimal: 
          1
          step: 
          1
          <button
            type="button"
          >
            laser-power
          </button>
        </div>
      </div>
      <div
        class="control"
      >
        <span
          class="label"
        >
          laser_speed
        </span>
        <div>
          Dummy Unit Input min: 
          3
          max: 
          300
          unit: 
          mm/s
          defaultValue: 
          5
          decimal: 
          1
          step: 
          1
          <button
            type="button"
          >
            laser-speed
          </button>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-11 css-dev-only-do-not-override-hash"
    >
      <div
        class="control"
      >
        <span
          class="label"
        >
          repeat
        </span>
        <div>
          Dummy Unit Input min: 
          1
          max: 
          100
          unit: 
          defaultValue: 
          5
          decimal: 
          0
          step: 
          1
          <button
            type="button"
          >
            laser-repeat
          </button>
        </div>
      </div>
      <div
        class="control"
      >
        <span
          class="label"
        >
          z_step
        </span>
        <div>
          Dummy Unit Input min: 
          0
          max: 
          20
          unit: 
          mm
          defaultValue: 
          5
          decimal: 
          2
          step: 
          0.5
          <button
            type="button"
          >
            laser-z-step
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
